pkg/awsutils/awsutils.go

```diff
@@ -838,33 +852,109 @@ func (cache *EC2InstanceMetadataCache) createENI(useCustomCfg bool, sg []*string
+
+			input.SubnetId = subnet.SubnetId
+			start := time.Now()
+			result, err = cache.ec2SVC.CreateNetworkInterfaceWithContext(context.Background(), input)
+			prometheusmetrics.Ec2ApiReq.WithLabelValues("CreateNetworkInterface").Inc()
+			prometheusmetrics.AwsAPILatency.WithLabelValues("CreateNetworkInterface", fmt.Sprint(err != nil), awsReqStatus(err)).Observe(msSince(start))
+			if err == nil {
+				log.Infof("Created a new ENI: %s", aws.StringValue(result.NetworkInterface.NetworkInterfaceId))
+				return *result.NetworkInterface.NetworkInterfaceId, nil
+			}
+			checkAPIErrorAndBroadcastEvent(err, "ec2:CreateNetworkInterface")
+			awsAPIErrInc("CreateNetworkInterface", err)
+			prometheusmetrics.Ec2ApiErr.WithLabelValues("CreateNetworkInterface").Inc()
+			log.Errorf("Failed to CreateNetworkInterface %v for subnet %s", err, *subnet.SubnetId)
+		}
 	}
+	return "", errors.Wrap(err, "failed to create network interface")
+}
 
-	log.Infof("Creating ENI with security groups: %v in subnet: %s", aws.StringValueSlice(input.Groups), aws.StringValue(input.SubnetId))
+func (cache *EC2InstanceMetadataCache) getVpcSubnets() ([]*ec2.Subnet, error) {
+	describeSubnetInput := &ec2.DescribeSubnetsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("vpc-id"),
+				Values: []*string{aws.String(cache.vpcID)},
+			},
+			{
+				Name:   aws.String("availability-zone"),
+				Values: []*string{aws.String(cache.availabilityZone)},
+			},
+		},
+	}
 
 	start := time.Now()
-	result, err := cache.ec2SVC.CreateNetworkInterfaceWithContext(context.Background(), input)
-	prometheusmetrics.Ec2ApiReq.WithLabelValues("CreateNetworkInterface").Inc()
-	prometheusmetrics.AwsAPILatency.WithLabelValues("CreateNetworkInterface", fmt.Sprint(err != nil), awsReqStatus(err)).Observe(msSince(start))
+	subnetResult, err := cache.ec2SVC.DescribeSubnetsWithContext(context.Background(), describeSubnetInput)
+	prometheusmetrics.Ec2ApiReq.WithLabelValues("DescribeSubnets").Inc()
+	prometheusmetrics.AwsAPILatency.WithLabelValues("DescribeSubnets", fmt.Sprint(err != nil), awsReqStatus(err)).Observe(msSince(start))
 	if err != nil {
-		checkAPIErrorAndBroadcastEvent(err, "ec2:CreateNetworkInterface")
-		awsAPIErrInc("CreateNetworkInterface", err)
-		prometheusmetrics.Ec2ApiErr.WithLabelValues("CreateNetworkInterface").Inc()
-		log.Errorf("Failed to CreateNetworkInterface %v", err)
-		return "", errors.Wrap(err, "failed to create network interface")
+		checkAPIErrorAndBroadcastEvent(err, "ec2:DescribeSubnets")
+		awsAPIErrInc("DescribeSubnets", err)
+		prometheusmetrics.Ec2ApiErr.WithLabelValues("DescribeSubnets").Inc()
+		return nil, errors.Wrap(err, "AllocENI: unable to describe subnets")
+	}
+
+	// Sort the subnet by available IP address counter (desc order) before determining subnet to use
+	sort.SliceStable(subnetResult.Subnets, func(i, j int) bool {
+		return *subnetResult.Subnets[j].AvailableIpAddressCount < *subnetResult.Subnets[i].AvailableIpAddressCount
```

Comment:
This might not work for PD completely..maybe a subnet has more available IPs but not /28 blocks i.e, more fragments..ideally for PD we should pick a subnet which has more /28 blocks i.e, contiguous 16IPs..

Reply:
I am thinking if we can retry on remaining subnets if we hit InsufficientCidr error...that should work for both PD and secondary IP mode..

Reply:
Agreed, that's a good plan

Reply:
Are we checking `InsufficientCidr` ?